Ed-Fi-Exchange-OSS/EdFi-Project-Buzz
# [FIF-299] Build Teamcity and create release in Octopus Deploy

## OVERVIEW

This is part of the task to create a TeamCity step that generates the Octopus Deployment release. 

As those commits will be inside the app install.ps1 scripts, we are instead moving that to a separate PR to get other work items unblocked.

## TEST

> **NOTE:** The following can only be tested in a main branch. We would need to commit to main before this can actually be tested. An experimental Team City build was created with a fork of this branch to test. I made you both developers on my private build so you can run the builds.

## TESTING STEPS TO GET READY TO INSTALL LOCALLY

- In EdFi-Project-Buzz\EdFi.Buzz.Installer\eng\Windows, copy example.configuration.json to configuration.json and alter the values for your desired local install configuration.
- Open a PowerShell window in Administrator mode and navigate to the EdFi-Project-Buzz root. $PSVersionTable.PSEdition should be Desktop.
- Execute .\test.ps1. This does a local build of all of the Buzz apps, and copies them to the C:\nugetRepo folder.

At this point, you should have a local repository of all nupkgs for the Buzz apps. Changing your configuration.json:artifactRepo setting to C:\nugetRepo will use this folder as your NuGet repository.

## TESTING THE INSTALLATION ON YOUR LOCAL DEVELOPMENT ENVIRONMENT

Given that you have a Core admin PowerShell console open, 
When you execute EdFi-Project-Buzz\EdFi.Buzz.Installer\eng\Windows\install.ps1, 
- It should throw an error that the script requires a non-Core PowerShell.

Given that you have a non-Core PowerShell console in non-admin role, 
When you execute EdFi-Project-Buzz\EdFi.Buzz.Installer\eng\Windows\install.ps1, 
- It should throw an error that the script requires administrator to execute.

Given that you have a non-core admin PowerShell and a valid configuration with a non-existent database, 
When you execute 
EdFi-Project-Buzz\EdFi.Buzz.Installer\eng\Windows\install.ps1, 
- It should install all apps to your configured folder (C:\Ed-Fi\Buzz by default), and 
- It should start all three services, and 
- It should create your new database, and 
- It should migrate the newly created database to the latest migration, and
- It should report the status of all installed services and webs (ETL, API and Web).

## TESTING ON THE EXPERIMENTAL BUILD

Given that you are on the [EXPERIMENTAL KJohnson Buzz Installer build](https://intedfitools1.msdf.org/project/Experimental_Kjohnson_Buzz_Buzz_Installer?mode=builds), 
When you click Run... for the Installer Branch Build,
- It should run the [Branch Build and Test](https://intedfitools1.msdf.org/buildConfiguration/Experimental_Kjohnson_Buzz_BranchInstallerBuild) for the installer successfully, and
- It should upload the version you just created edfi.buzz.installer to [Azure Artifacts](https://dev.azure.com/ed-fi-alliance/Ed-Fi-Alliance-OSS/_packaging?_a=feed&feed=EdFi) and
- It should then start the [Installer Deploy build configuration](https://intedfitools1.msdf.org/buildConfiguration/Experimental_Kjohnson_Buzz_DeployInstallerBuild) and use the package version number that was created during in the build process, and
- The Installer Deploy build should succeed, and 
- It should [generate a Release on Octopus](https://intedfideploy1.msdf.org/app#/projects/project-buzz/overview)
- Octopus Deploy should execute the Installer package version number that was created in the Installer build process.

### NOTES
The Branch Build and Test build configuration has a `Step 2 - Package (PowerShell)` that will output the octopus.release.version. This value should be used to create the Octopus Release version in the Deploy build configuration.
